src/lakefs_spec/spec.py

```diff
@@ -422,14 +440,33 @@ def ls(self, path: str | os.PathLike[str], detail: bool = True, **kwargs: Any) -
         if info:
             # assumes that the returned info is name-sorted.
             pp = self._parent(info[0]["name"])
+            info_copy = info[:]
             if pp in self.dircache:
                 # ls info has files not in cache, so we update them in the cache entry.
-                cache_entry = self.dircache[pp]
-                # extend the entry by the new ls results
-                cache_entry.extend(info)
+                cache_entry = self.dircache[pp][:]
+
+                old_names = {e["name"] for e in cache_entry}
+                new_names = {e["name"] for e in info_copy}
+
+                to_remove = old_names - new_names
+                to_update = old_names.intersection(new_names)
+
+                # Remove all entries no longer present in the current listing
+                cache_entry = [e for e in cache_entry if e["name"] not in to_remove]
+
+                # Overwrite existing entries in the cache with its updated values
+                for name in to_update:
+                    old_idx = next(idx for idx, e in enumerate(cache_entry) if e["name"] == name)
+                    new_entry = next(e for e in info_copy if e["name"] == name)
+
+                    cache_entry[old_idx] = new_entry
+                    info_copy.remove(new_entry)
+
+                # Add the remaining (new) entries to the cache
+                cache_entry.extend(info_copy)
                 self.dircache[pp] = sorted(cache_entry, key=operator.itemgetter("name"))
             else:
-                self.dircache[pp] = info
+                self.dircache[pp] = info[:]
```

Comment:
I'm unsure about the behavior here for subdirectories if `recursive=True`: On the one hand, they are contained in the parent, so their contents could go in the parent cache entry just fine.

On the other hand, they are also directories, so they could deserve their own cache entry.

We could just save the directory stub in the cache entry in the cache entry, and then materialize it by more cache lookups if `recursive=True`:

```py
data/a.txt
       /b.txt
       /dir/f.txt

-> fs.ls("data") = [{"name": "a.txt", ...}, {"name": "b.txt", ...}, {"name": "dir", ..., type="directory"}]
```
and then looking up `dir` in `fs.dircache` if `recursive=True`.

Reply:
I think this goes in the direction of my forgotten debugging statement in the original PR, which was collecting all the directories that were part of the `list_objects` response.

We currently have this, which I believe is not enough:

```python
            # assumes that the returned info is name-sorted.
            pp = self._parent(info[0]["name"])
```

Rather, we need to add every parent directory contained in the response to the cache, as you mentioned above. That's an easy enough fix, but it should be accompanied by a corresponding test.

Reply:
Let's follow up separately on this one to unblock the merge of this PR.